org-roam-db.el

```diff
@@ -280,26 +231,56 @@ If UPDATE-P is non-nil, first remove the meta for the file in the database."
     (org-roam-db-query
      [:insert :into files
       :values $v1]
-     (list (vector file hash (list :atime atime :mtime mtime))))))
+     (list (vector file hash atime mtime)))))
 
-(defun org-roam-db--insert-titles (&optional update-p)
-  "Update the titles of the current buffer into the cache.
-If UPDATE-P is non-nil, first remove titles for the file in the database.
-Returns the number of rows inserted."
-  (let* ((file (or org-roam-file-name (buffer-file-name)))
-         (titles (or (org-roam--extract-titles)
-                     (list (org-roam--path-to-slug file))))
-         (rows (mapcar (lambda (title)
-                         (vector file title)) titles)))
+(defun org-roam-db-insert-nodes (&optional update-p)
+  (let ((file (or org-roam-file-name
+                  (buffer-file-name (buffer-base-buffer))))
+        nodes
+        id level pos tags title ref)
     (when update-p
-      (org-roam-db-query [:delete :from titles
+      (org-roam-db-query [:delete :from nodes
                           :where (= file $s1)]
                          file))
-    (org-roam-db-query
-        [:insert :into titles
-         :values $v1]
-        rows)
-    (length rows)))
+    (org-with-point-at 1
+      ;; First, we get the file-level ID
+      (when (setq id (org-id-get))
+        (setq title (cadr (assoc "TITLE" (org-collect-keywords '("title"))
+                                 #'string-equal))
+              ;; TODO handle tags
+              tags nil
+              pos (point)
+              level (org-outline-level)
+              ;; TODO handle ref
+              ref nil)
+        (push (vector id file level pos tags title ref) nodes))
+      ;; Then we loop over all headlines
+      (org-map-entries
+       (lambda ()
+         (when (setq id (org-id-get))
+           (setq title (nth 4 (org-heading-components))
+              ;; TODO handle tags
+              tags nil
+              pos (point)
+              level (org-outline-level)
```

Comment:
If you are using `org-heading-components`, you can get level from there, saves some parsing. Otherwise, instead of `org-heading-components` you can use `(org-entry-get nil "ITEM")`.